### PR TITLE
Add option for Operator to pass arguments to GrafanaAgent #1227

### PR DESCRIPTION
**Pull request by shamsalmon**
#### PR Description 

Adds Args CRD to agent. Mostly for new config read endpoint options (disabled by default).

#### Which issue(s) this PR fixes 

https://github.com/grafana/agent/issues/1227

#### Notes to the Reviewer

Can be enabled like so:
```
apiVersion: monitoring.grafana.com/v1alpha1
kind: GrafanaAgent
metadata:
  name: grafana-agent
  namespace: operations
  labels:
    app: grafana-agent
spec:
  image: grafana/agent:v0.21.2
  logLevel: info
  serviceAccountName: grafana-agent
  args:
    - --config.enable-read-api
  metrics:
    instanceSelector:
      matchLabels:
        agent: grafana-agent-metrics
    externalLabels:
      cluster: devqa
```

#### PR Checklist

- [x] CHANGELOG updated 
- [ ] Documentation added
- [ ] Tests updated

### CODE REVIEW

**Comment by shamsalmon** [comment]
Also https://github.com/grafana/helm-charts/pull/951 for helm support for this new feature

File: pkg/operator/apis/monitoring/v1alpha1/types.go
Comment (rlankfo) [inline pkg/operator/apis/monitoring/v1alpha1/types.go:152]:
```suggestion
	// +kubebuilder:default={"-config.file=/var/lib/grafana-agent/config/agent.yml","-config.expand-env=true","-reload-port=8081"}
	Args []string `json:"args,omitempty"`
```

**Review by rlankfo — APPROVED** [review]
LGTM :+1: but I would suggest using the default marker to move the arguments still defined in code to the default value for args as demonstrated in review. This will hopefully prevent confusion if a user wants to use this args field to set any of the values that are still currently hardcoded. 

This should allow us to get rid of `agentArgs` slice and pass `d.Agent.Spec.Args` on line 403. Note you'll need to re-run `make crds`.

File: production/operator/crds/monitoring.grafana.com_grafanaagents.yaml
Comment (rlankfo) [inline production/operator/crds/monitoring.grafana.com_grafanaagents.yaml:1068]:
```suggestion
                description: ConfigMaps is a list of config maps in the same namespace
```

This is unrelated to your change but I just noticed this typo

**Comment by rfratto** [comment]
Thanks for the PR @shamsalmon!

I'm a little concerned about the idea of generally exposing the container args: 

* We might not want to allow full-override of the arguments, since the reload port *must* always be on 8081, otherwise things will silently break.
* On the other hand, we might not want to allow appending to the arguments, since providing an already-defined argument would also cause things to break (though it would break more noisily with container restart loops). 

I also _think_ it's already possible to use the `containers` field of the CRD to fully override the container arguments. That field was copied from the Prometheus Operator, and it's documented as explicitly an advanced option that has no backwards compatibility guarantees between releases. 

My gut feeling is:

1. This new option should append to arguments rather than be a full replace (to supplement the existing merge functionality), and it should be similarly documented as an advanced option with no backwards compatibility guarantees.
2. There should also be a new boolean option in the CRD specifically for enabling `enableConfigReadAPI` which will automatically provide the argument.

WDYT @rlankfo?

**Comment by rfratto** [comment]
> I'm fine with adding a new option in the CRD for a specific flag, but that sets a precedent of needing to do this for every flag we may want to expose which would be more maintenance.

Right, I at least think enabling the read API is something should explicitly expose, but I agree we can't be expected to have a CRD option for literally every flag we might have. The operator is intended to be a Kubernetes-tailored experience, which isn't necessarily conducive to have a 1:1 mapping of flags.

**Comment by rlankfo** [comment]
@rfratto I'll spend some time to verify if our args can be overridden using the `containers` field or if this just gets merged with the existing args we have hardcoded.

In any case, it sounds like the path forward here is probably just to enable the boolean for `enableConfigReadAPI`. I could see us wanting to use remote config here (one reason I wanted to be able to add args for experimental flags, and for modifying -config.file) but this could also be an explicit field.

Personally I think we should avoid appending user defined args with hardcoded args but I'm fine if that ends up being the direction we want to take.

Let me know what you think.

**Comment by rfratto** [comment]
I agree that only adding an extra option for enabling the config read API is the least contentious path for now. We can revisit a general option for appending new options later if this comes up again, but we aren't adding many options that often.

@shamsalmon sorry for the back and forth, but would you mind changing your PR so that the only new field to the CRD is for explicitly enabling the config read API?

**Comment by shamsalmon** [comment]
> I agree that only adding an extra option for enabling the config read API is the least contentious path for now. We can revisit a general option for appending new options later if this comes up again, but we aren't adding many options that often.
> 
> @shamsalmon sorry for the back and forth, but would you mind changing your PR so that the only new field to the CRD is for explicitly enabling the config read API?

will do!

**Comment by shamsalmon** [comment]
> Hey @shamsalmon, I just wanted to check in on this PR. Is there anything we can do to help out on this? Thanks.

Sorry for the delay. I got busy with work and forgot about it! I will get the change pushed out today / tomorrow.

**Comment by shamsalmon** [comment]
Just wrapped up my changes and tested them out. Let me know if you have any other suggestions. I tested with the param enabled, with it disabled, and with it missing and it worked as expected.

**Review by rfratto — APPROVED** [review]
Hey @shamsalmon, sorry for the delay! LGTM, thanks for contributing!

File: CHANGELOG.md
Comment (rfratto) [inline CHANGELOG.md:7]: Can you rebase on top of current main and move this to the unreleased section?

**Review by rfratto — COMMENTED** [review]
Before we merge, can you sign the CLA?

**Comment by shamsalmon** [comment]
> Before we merge, can you sign the CLA?

I signed the CLA but for some reason its still showing as not signed. Any ideas?

**Comment by rfratto** [comment]
> I signed the CLA but for some reason its still showing as not signed. Any ideas?

It looks like the CLA bot is thrown off by the committer being sshannon, which doesn't match your GitHub username (& there's no email address to fall back on)

**Comment by rfratto** [comment]
If it's too much work, I can re-open this PR under my account and give you authorship credit :)

**Comment by shamsalmon** [comment]
> If it's too much work, I can re-open this PR under my account and give you authorship credit :)

Looks like I have to rewrite history to change my username etc. Is that ok or would you rather me open a new PR? Also ok if you would like to do it.

File: CHANGELOG.md
Comment (shamsalmon) [reply CHANGELOG.md]: done

**Comment by rfratto** [comment]
> Looks like I have to rewrite history to change my username etc. Is that ok or would you rather me open a new PR? Also ok if you would like to do it.
>
> Edit - Tried rewiritng history - looks like it worked. Sorry I somehow seem to have duplicated a bunch of writes.

rewriting history is fine 👍 I set this to automerge once the CI goes green. Thanks again for the contribution!

**Pull request merged**
